sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs

```diff
@@ -2014,6 +2015,57 @@ impl ty::TyExpression {
         }
     }
 
+    fn type_check_ref(
+        handler: &Handler,
+        mut ctx: TypeCheckContext<'_>,
+        expr: Box<Expression>,
+        span: Span,
+    ) -> Result<ty::TyExpression, ErrorEmitted> {
+        return match expr.kind {
+            ExpressionKind::Break | ExpressionKind::Continue | ExpressionKind::Return(_) => Err(
+                handler.emit_err(CompileError::ExpressionCannotBeReferenced {
+                    expression: NonReferenceableExpression::from(expr.kind),
+                    span,
+                }),
+            ),
+            _ => {
+                let engines = ctx.engines();
+                let type_engine = ctx.engines().te();
+
+                let ctx = ctx
+                    .by_ref()
+                    .with_type_annotation(type_engine.insert(engines, TypeInfo::Unknown, None))
```

Comment:
The type is really unknown here? I  think `TypeContext` may have some tips... this generates problems with `u8` an u64`, for example:

```
let a: &u8 = &1
```

Or something like this, because `1` will be treated as `numeric` until it decays to `u64` sometime later, creating an invalid expression.

Reply:
Good question, assuming that the initial issues, were exactly in the area of type inference and unification. Here is a detailed answer.

Well, `TypeInfo::Unknown` is actually our way of saying `without_type_annotation` ;-). the `type_annotation` TypeId is not an `Option` and setting it to `Unknown` is a way of saying we are fine with any type, no type expectations.This is how type annotation removal is done in all places, and now when I think about it, I would love to have an explicit `without_type_annotation` method to give a clear intent.

And yes in this case we want to remove the expectation that we are getting because if we have one, like in your example where the surrounding context will expect `&8` the expectation will always be for a reference type `&Something`. And in the type check we are actually checking the referenced expression `Something` which will never be `&Something`.

The returned type like `Numeric` in your example is packed into `Ref` and returned, and the later unification in the context (`let` in the above case) does proper unification.

In other words, no worry, the given example works :-)

BDW, since unification issues were so prominent in the beginning I've decided to have `unifying_reference_types` tests but didn't write them because to a good extent unification is implicitly covered in other tests. But the main reason was honestly that I got tired of writing tests. Nevertheless, the `unifying_reference_types` will come in upcoming PRs. Until then I trust my smoke tests, that also have examples like yours :-)

P.S. When implementing `impl<T> &T` or `impl<T> MyTrait for &T` we still get unification issues in some cases. But these features are at the moment considered as not implemented.

Reply:
@xunilrj It turned out that you were right. The above example indeed work, but there were issues in the type inference coming from the fact that `TypeInfo::Unknown` was used. #5643 is addressing those issues.